### ISSUE TITLE
Option to disable composer

### ISSUE THREAD

**Issue by olavmrk** [cloudfoundry/php-buildpack#74]
If I am reading the code corectly, Composer build is triggered if it happens to find a `composer.json` or `composer.lock` anywhere in the code that is pushed: https://github.com/cloudfoundry/php-buildpack/blob/v3.2.1/extensions/composer/extension.py#L154

Sometimes you do not want the buildpack to run composer for you, but would rather run it yourself before pushing the buildpack. (E.g. because you need access to some specific repository using your SSH key.)

It appears that the only way to disable composer currently is to delete all `composer.json` and `composer.lock` files, deploy the app, and then restore the composer files afterwards.

It would be nice if there was a better way to disable composer (an option in `.bp-config/options.json`?).

**Comment by dmikusa-pivotal** [comment]
I would like to see Composer only look for a top level composer.json|lock file.  I don't like the current behavior where it searches recursively for them.  If you vendor a library that has a composer.json|lock file, the bp will currently find it and try to run Composer on it, which the bp shouldn't do.

If we fixed that behavior, would that be sufficient?  Or do you push applications that have a top level composer.json|lock file (i.e. one for your app) where you don't want Composer to run?  Also, what's the impact of it running?  My understanding is that if you vendor stuff, Composer will run pretty quick, do nothing and exit.  Are you seeing something else?

Thanks!

**Comment by olavmrk** [comment]
I have fought a couple of battles with the Composer integration:

Today it was the case you mentioned -- it found a composer.json in a subdirectory, decided that it shoud run composer, and failed, since there was no top-level composer.json.

Earlier I deployed Piwik via the PHP buildpack. Piwik runs a sanity check on its files during the installation process, (probably to make sure that the user hasn't mangled the files during transfer). After this buildpack runs composer, it deletes all composer.json files for some reason. That caused Piwik to complain, because from its point of view, there was files missing. (I had to fix that by patching the Piwik manifest.)

A colleague had the problem I mentioned in the ticket -- he was deploying composer plugins that are only available on our internal soruce repositories, and ran into some trouble when the buildpack tried to run composer. I think he mentioned that he solved it by deleting `composer.json` before pushing it, but I don't have all the details.

I feel like the simplest way to solve this would be to just make it possible to disable composer. Or even better: a flag to disable all special operations taken during installation. Sometimes all you need is a simple way to just push a PHP application "as-is", and have it start up a web server with a specific web root.

**Comment by dmikusa-pivotal** [comment]
>Today it was the case you mentioned -- it found a composer.json in a subdirectory, decided that it shoud run composer, and failed, since there was no top-level composer.json.

Ok, this is definitely something we need to fix.  In the meantime, you can workaround it by adding a minimal composer.json file in your project.

Ex:

```
{}
```

>Earlier I deployed Piwik via the PHP buildpack. Piwik runs a sanity check on its files during the installation process, (probably to make sure that the user hasn't mangled the files during transfer). After this buildpack runs composer, it deletes all composer.json files for some reason. That caused Piwik to complain, because from its point of view, there was files missing. (I had to fix that by patching the Piwik manifest.) 

The build pack doesn't delete them, but it does move certain files around.  It's likely that they're just not where Piwik expects them to be.  I haven't used that project so I can't say for sure, but it's likely that you could work around this by either creating an `htdocs` directory in the root of your project and putting the Piwik files in that or by setting `WEBDIR` in `options.json` to an existing directory.

The reason that the build pack moves files is because when you upload an application directly it doesn't know what to set as the docroot.  It does it's best to setup a docroot and move your files there, but in some cases this can cause problems.  If you manually create the default docroot, which is `htdocs` or you set the docroot, done with `WEBDIR`, to an existing folder the build pack won't need to create a docroot and move files around.


>A colleague had the problem I mentioned in the ticket -- he was deploying composer plugins that are only available on our internal soruce repositories, and ran into some trouble when the buildpack tried to run composer. I think he mentioned that he solved it by deleting composer.json before pushing it, but I don't have all the details.

That sounds like it might work or possibly just add an minimal composer.json file.  Another option would be to vendor the files.  When Composer runs, it should check for a local copy first and not download them if the local copy exists.  


>I feel like the simplest way to solve this would be to just make it possible to disable composer.

Technically this is an option.  I just want to make sure we're not masking over other problems.  I'd prefer to address the real issues rather than just making convenient workarounds.


>Or even better: a flag to disable all special operations taken during installation. Sometimes all you need is a simple way to just push a PHP application "as-is", and have it start up a web server with a specific web root.

See my comment above about the docroot.  It won't disable Composer, but it should do what you're asking here and will result in the simplest setup.

**Comment by olavmrk** [comment]
> > Today it was the case you mentioned -- it found a composer.json in a subdirectory, decided that it shoud run composer, and failed, since there was no top-level composer.json.
> 
> Ok, this is definitely something we need to fix. In the meantime, you can workaround it by adding a minimal composer.json file in your project.
> 
> Ex:
> 
>   {}

Well, at least it is a cleaner hack than:

```sh
find "$SSP_DIR" '(' -name 'composer.json' -o -name 'composer.lock' ')' -delete
```

> > Earlier I deployed Piwik via the PHP buildpack. Piwik runs a sanity check on its files during the installation process, (probably to make sure that the user hasn't mangled the files during transfer). After this buildpack runs composer, it deletes all composer.json files for some reason. That caused Piwik to complain, because from its point of view, there was files missing. (I had to fix that by patching the Piwik manifest.)
> 
> The build pack doesn't delete them, but it does move certain files around. It's likely that they're just not where Piwik expects them to be. I haven't used that project so I can't say for sure, but it's likely that you could work around this by either creating an htdocs directory in the root of your project and putting the Piwik files in that or by setting WEBDIR in options.json to an existing directory.

This is the way I deploy it. I extract the package to the `piwik`-directory and set `WEBDIR` to `piwik`. Looking at the code now, I assume that this part is responsible:

https://github.com/cloudfoundry/php-buildpack/blob/v3.2.1/extensions/composer/extension.py#L253

> > A colleague had the problem I mentioned in the ticket -- he was deploying composer plugins that are only available on our internal soruce repositories, and ran into some trouble when the buildpack tried to run composer. I think he mentioned that he solved it by deleting composer.json before pushing it, but I don't have all the details.
> 
> That sounds like it might work or possibly just add an minimal composer.json file. Another option would be to vendor the files. When Composer runs, it should check for a local copy first and not download them if the local copy exists.

I had a closer look at this now, and think I have found the reason it failed: This package uses `lib/vendor/` instead of the default `vendor/`. When one naively runs `composer.phar install`, it stores everything in `vendor/`, which then is ignored when the app is pushed. When I instead ran:

```
COMPOSER_VENDOR_DIR=lib/vendor composer.phar install
```

it worked nicely. Composer found the existing libraries, and it worked.

> > I feel like the simplest way to solve this would be to just make it possible to disable composer.
> 
> Technically this is an option. I just want to make sure we're not masking over other problems. I'd prefer to address the real issues rather than just making convenient workarounds.

Well, thinking about it, I guess a large part of wanting such an option is frustration on my part. I feel like I have repeatedly fought with the composer-functionality in this package in order to get things to work. I know it tries to be helpful, but sometimes it just gets in the way.

I think the option you mentioned earlier -- making it only look at top-level `composer.json` -- would mostly work for my part. At least until I run into something that causes `composer install` to fail even then :)

Combine that with a fix so that it doesn't move around on `composer.json` and `composer.lock` in `WEBDIR`, and I guess I should be able to adjust `.bp-config/options.json` to work with most of my applications.

**Comment by dmikusa-pivotal** [comment]
>>>Earlier I deployed Piwik via the PHP buildpack. Piwik runs a sanity check on its files during the installation process, (probably to make sure that the user hasn't mangled the files during transfer). After this buildpack runs composer, it deletes all composer.json files for some reason. That caused Piwik to complain, because from its point of view, there was files missing. (I had to fix that by patching the Piwik manifest.)

>>The build pack doesn't delete them, but it does move certain files around. It's likely that they're just not where Piwik expects them to be. I haven't used that project so I can't say for sure, but it's likely that you could work around this by either creating an htdocs directory in the root of your project and putting the Piwik files in that or by setting WEBDIR in options.json to an existing directory.

>This is the way I deploy it. I extract the package to the piwik-directory and set WEBDIR to piwik. Looking at the code now, I assume that this part is responsible:

>https://github.com/cloudfoundry/php-buildpack/blob/v3.2.1/extensions/composer/extension.py#L253

This is intentional.  The build pack moves these two files because they would otherwise be publicly accessible.  It's not the end of the world if they are public, but they do contain details like library names and versions being used by the app.  The build pack wants to error on the side of caution and tries to not expose this information, which could provide clues for someone wanting to attack the application.

**Comment by dmikusa-pivotal** [comment]
>>>A colleague had the problem I mentioned in the ticket -- he was deploying composer plugins that are only available on our internal soruce repositories, and ran into some trouble when the buildpack tried to run composer. I think he mentioned that he solved it by deleting composer.json before pushing it, but I don't have all the details.
>>That sounds like it might work or possibly just add an minimal composer.json file. Another option would be to vendor the files. When Composer runs, it should check for a local copy first and not download them if the local copy exists.
>I had a closer look at this now, and think I have found the reason it failed: This package uses lib/vendor/ instead of the default vendor/. When one naively runs composer.phar install, it stores everything in vendor/, which then is ignored when the app is pushed. When I instead ran:

>COMPOSER_VENDOR_DIR=lib/vendor composer.phar install
>it worked nicely. Composer found the existing libraries, and it worked.

The build pack by default puts the vendor directory at `$HOME/lib/vendor`.  It does this because the lib directory is on PHP's include path, the build pack puts it there, and so it makes it easy to access libraries installed by Composer.

Your solution is perfect.  Another option is to set COMPOSER_VENDOR_DIR in `options.json`.  If your app needs the vendor directory in another location, you can use that option to move it.

**Comment by dmikusa-pivotal** [comment]
>Combine that with a fix so that it doesn't move around on composer.json and composer.lock in WEBDIR, and I guess I should be able to adjust .bp-config/options.json to work with most of my applications.

I suppose we could make this an option. It's tricky though, for the reason I mentioned previously.  If we did make it an option, I'd want the default behavior to be as it is now.  In other words, move those files out of the docroot.  That would then leave users able to opt-out if they're OK with having those files public.

**Comment by dmikusa-pivotal** [comment]
Take aways from this:

1.) We need to fix the way the BP searches for composer.json.  I think it should look for it in two places, BUILD_DIR/composer.json and BUILD_DIR/WEBDIR/composer.json.  If it's not found there, it would be up to the user to set the COMPOSER environment variable to point Composer to the file.

   https://getcomposer.org/doc/03-cli.md#composer

2.) Add option to allow user to opt out of moving composer.* files.  Alternatively, look at some other way of securing these files, which should not be publicly available.

**Comment by olavmrk** [comment]
> > This is the way I deploy it. I extract the package to the piwik-directory and set WEBDIR to piwik. Looking at the code now, I assume that this part is responsible:
> >
> >   https://github.com/cloudfoundry/php-buildpack/blob/v3.2.1/extensions/composer/extension.py#L253
> 
> This is intentional. The build pack moves these two files because they would otherwise be publicly accessible. It's not the end of the world if they are public, but they do contain details like library names and versions being used by the app. The build pack wants to error on the side of caution and tries to not expose this information, which could provide clues for someone wanting to attack the application.

I'd say that if they are deliberately put in `WEBDIR` (where the user has manually created a `WEBDIR`-directory and set the `WEBDIR` option), they should be safe to publish. I at least do not put files in `WEBDIR` that should be secret :smiley:

Btw.: What would happen here if I also had a `composer.json` in `BUILD_DIR`? If I read the code correctly, it would overwrite my `composer.json` with whatever was in `WEBDIR`. That would be very confusing if it ever happened.

**Comment by olavmrk** [comment]
> Take aways from this:
> 
> 1.) We need to fix the way the BP searches for composer.json. I think it should look for it in two places, BUILD_DIR/composer.json and BUILD_DIR/WEBDIR/composer.json. If it's not found there, it would be up to the user to set the COMPOSER environment variable to point Composer to the file.
> 
> https://getcomposer.org/doc/03-cli.md#composer
> 
> 2.) Add option to allow user to opt out of moving composer.* files. Alternatively, look at some other way of securing these files, which should not be publicly available.

A "counterproposal" :smiley: :

Only look in `BUILD_DIR` for `composer.json`. If the user hasn't specified a `WEBDIR`, move files that look like they belong there into a autogenerated `WEBDIR` (I get the impression that you already do this?). Do not move the `composer.json` and `composer.lock` files in that case.

The result is that only the top-level `composer.json` is executed, and it will never be published.

That at least would be the behavior that I find most intuitive. A `WEBDIR` is after all something that one creates in order to separate the files reachable through the webserver away from library and configuration files. I do not think users have an expectation of files stored in `WEBDIR` to not be reachable through the web server. If they put sensitive files in `WEBDIR`, they are most likely already in the habit of protecting them through some other means.

**Comment by dmikusa-pivotal** [comment]
>I'd say that if they are deliberately put in WEBDIR (where the user has manually created a WEBDIR-directory and set the WEBDIR option), they should be safe to publish. I at least do not put files in WEBDIR that should be secret :smiley:

Fair enough.  I'm in favor of trying to make things just work, but ultimately the user should have the control, even if that provides the opportunity to shoot one's self in the foot.

>Btw.: What would happen here if I also had a composer.json in BUILD_DIR? If I read the code correctly, it would overwrite my composer.json with whatever was in WEBDIR. That would be very confusing if it ever happened.

I'm not sure exactly, I'd have to test.  As a guess, I think it would detect the composer.json in build_dir, since it searches top down and just ignore the one in WEBDIR.

**Comment by dmikusa-pivotal** [comment]
>Only look in BUILD_DIR for composer.json. If the user hasn't specified a WEBDIR, move files that look like they belong there into a autogenerated WEBDIR (I get the impression that you already do this?). Do not move the composer.json and composer.lock files in that case.

This would technically be tricky because of the way the build pack is structured.  It works with extensions, so all of the Composer behavior is isolated into that extension.  The problem is that if no docroot is set, files are moved into one prior to running any extensions.

It leaves us with the following choices:
  - Have extensions move files they know about back out of the docroot
  - Change when the move occurs, which would likely affect a lot of the logic of the build pack since there are assumptions made about when this occurs and where files will exist.
  - Add knowledge about Composer outside of the extension.

It's also tricky to tell what "should" be in the docroot.  Right now we move everything, except a few files and directories.  It would be easy to add the composer.json|lock files to this list, but you're putting knowledge of Composer outside the extension.  It would probably be better in the log run to modify the extensions to have a say in what files are and are not moved.

https://github.com/cloudfoundry/php-buildpack/blob/master/lib/compile_helpers.py#L45-L56

>The result is that only the top-level composer.json is executed, and it will never be published.

Based on other issues that have come up, I know that we have users who put everything in the docroot and run from there.  I don't want to break that use case.  That said, if we found the composer.json file there, we'd move it to the build_dir, so either way Composer would always run from the top level.

>That at least would be the behavior that I find most intuitive. A WEBDIR is after all something that one creates in order to separate the files reachable through the webserver away from library and configuration files. 

I agree, but not all PHP apps are structured this way.  It's one of the challenges we hit a lot on the build pack.

>I do not think users have an expectation of files stored in WEBDIR to not be reachable through the web server. If they put sensitive files in WEBDIR, they are most likely already in the habit of protecting them through some other means.

I agree, but I don't think everyone thinks about this.  I think the build pack should do what it can within reason to protect any files it knows should not be public.  Currently it does this with the help of the file system, i.e. it moves things out of the docroot that it knows should not be public.  I favor this approach because I find it to the be safest.  You can protect files with the web server but bugs and configuration mistakes, mostly the latter, happen and so I don't like that approach as much.

**Comment by olavmrk** [comment]
> * Have extensions move files they know about back out of the docroot

Isn't this more or less what happens now? The composer extension moves `composer.json` and `composer.lock` out of `WEBDIR` and into `BUILD_DIR`?

To implement a method this, I think it should be enough to give the composer extension information about whether the `WEBDIR` as something the user asked for or if it was something that was autogenerated. This information can be provided as general data to extensions, so it isn't composer specific.

The composer extension can then use that data in order to decide whether to move those files out of `WEBDIR` to `BUILD_DIR` and run composer on them.

> It's also tricky to tell what "should" be in the docroot. Right now we move everything, except a few files and directories. It would be easy to add the composer.json|lock files to this list, but you're putting knowledge of Composer outside the extension. It would probably be better in the log run to modify the extensions to have a say in what files are and are not moved.

That is probably a better long term solution. A simple extension point that allows each extension a chance to move files it doesn't think belongs in a autogenerated `WEBDIR` back to `BUILD_DIR`.

> > That at least would be the behavior that I find most intuitive. A WEBDIR is after all something that one creates in order to separate the files reachable through the webserver away from library and configuration files.
>
> I agree, but not all PHP apps are structured this way. It's one of the challenges we hit a lot on the build pack.

Yes, but those that aren't built this way already has mechanisms to deal with files being available in the document root, through various means. See for example how Piwik ensures that its configuration files can't be downloaded:

https://github.com/piwik/piwik/blob/2.13.1/config/global.ini.php#L1

> > I do not think users have an expectation of files stored in WEBDIR to not be reachable through the web server. If they put sensitive files in WEBDIR, they are most likely already in the habit of protecting them through some other means.
>
> I agree, but I don't think everyone thinks about this. I think the build pack should do what it can within reason to protect any files it knows should not be public. Currently it does this with the help of the file system, i.e. it moves things out of the docroot that it knows should not be public. I favor this approach because I find it to the be safest. You can protect files with the web server but bugs and configuration mistakes, mostly the latter, happen and so I don't like that approach as much.

But does the buildpack need to try to solve a problem that PHP has already dealt with for a number of years? Or maybe a far more relevant question: Will anybody complain that this buildpack didn't protect them when they had specified a `WEBDIR`, but accidentally had left a sensitive file in it? No matter how hard you try, there will always be room for users making mistakes. Take for example the quite common development pattern:

```
cp config.php config.php.bak
$EDITOR config.php
```

There is no way that you can protect against all variants of that, and it is far more likely that someone will accidentally publish sensitive information that way than by creating a `WEBDIR`, adding a `composer.json` to it, and leaving something sensitive in the `composer.json`.

**Comment by dmikusa-pivotal** [comment]
>>Have extensions move files they know about back out of the docroot
>Isn't this more or less what happens now? The composer extension moves composer.json and composer.lock out of WEBDIR and into BUILD_DIR?

Yes

>Yes, but those that aren't built this way already has mechanisms to deal with files being available in the document root, through various means. See for example how Piwik ensures that its configuration files can't be downloaded:
>https://github.com/piwik/piwik/blob/2.13.1/config/global.ini.php#L1

Interesting.  I haven't seen that approach before.  Seems like it should work fine with the build pack though.

>But does the buildpack need to try to solve a problem that PHP has already dealt with for a number of years?

The build pack is responsible for configuring a secure environment, which ultimately means configuring the web server & PHP in a secure way.  We do this by trying to follow the generally accepted practices of the community.  I think what we've done likely uses, builds on or works with the things that you're talking about.

>Or maybe a far more relevant question: Will anybody complain that this buildpack didn't protect them when they had specified a WEBDIR, but accidentally had left a sensitive file in it? No matter how hard you try, there will always be room for users making mistakes.

I think it's reasonable for a user to assume that the build pack is deploying their app is a safe and secure manner.  It's not magic though, as you said there will be the potential for a user to do something that side-steps security and results in a problem.

**Comment by flavorjones** [comment]
Great conversation, guys.

@dmikusa-pivotal: To make sure I haven't missed anything, could you summarize the actions you recommend we take as a result of this Issue?

**Comment by dmikusa-pivotal** [comment]
Summary:

1.) We need to fix the way the BP searches for composer.json. I think it should look for it in two places, BUILD_DIR/composer.json and BUILD_DIR/WEBDIR/composer.json. If it's not found there, it would be up to the user to set the COMPOSER environment variable to point Composer to the file.  We should document that option.

https://getcomposer.org/doc/03-cli.md#composer

2.) Add option to allow user to opt out of moving composer.* files (the default would be the current behavior, which is to move them).  This would probably require us to set the COMPOSER environment variable when we run composer so that it finds the correct composer.json file, or to run composer from a different directory.  I think the env variable approach would probably be easier.  It would also either require us to secure the composer files in a different way, perhaps with web server config, or print a WARNING message to the user that the files are in a public location and leave that up to the user.

As an aside, it wouldn't hurt to continue doing our current approach of moving files out of the docroot to secure them and at the same time add config to the web servers to restrict access to those files.  It shouldn't hurt anything to protect on two levels.  Plus if the user opts out of moving composer.*, it would still leave them with some protection.

**Comment by olavmrk** [comment]
Hi,

sorry for the lack of follow-up from me these last days -- I have been traveling and working on other stuff these last days, and am just now staring to get caught up on email.

I do think those two options would make my life easier, and they will certainly solve the issues I have had. I am slightly worried about the number of options I will have to tune in order to ensure that the buildpack doesn't do anything too imaginative / special, but once I figure it out I guess I will just copy them between different sites, so it will not be a big problem.

**Comment by flavorjones** [comment]
The team will discuss this week at our iteration planning meeting.

**Comment by flavorjones** [comment]
We're going to stop looking for composer.json files anywhere in the hierarchy. This work is split out under a new story:

> https://www.pivotaltracker.com/story/show/99170278

**Comment by jornane** [comment]
I'm from the same company as @olavmrk. Lately, i've been trying to get the Piwik web analysis software to install on Cloudfoundry. The way Piwik works, is that the `piwik` folder you download contains `vendor`, but also `index.php`, making it the webroot.

Piwik internals will expect `vendor` in the same directory as `index.php`, but the build pack moves it away. Piwik has this path hardcoded and will not try include paths if it doesn't find `vendor` where it is expecting it.

Since Piwik comes with a prepopulated `vendor` folder which does not require that composer be run, I think an option to disable composer would be the best solution here. Is there any update on the issue?

**Comment by dmikusa-pivotal** [comment]
With the fix that @flavorjones linked to (currently only on the develop branch), you can effectively disable Composer by adding `.cfignore` file in the root directory of your project (same as `manifest.yml`) and listing in it the `composer.json` and `composer.lock` files (syntax is the same as `.gitignore`).  

That will prevent those files from being uploaded by `cf` and if they don't exist on the server then Composer won't be run.  

Is that a sufficient way to disable it for your needs?

**Comment by Dannyzen** [comment]
Closing this out as this thread is > 4 months old. Please re-open this issue if additional assistance is needed.
